lib/hardwarecal/hardwarecal.go

```diff
@@ -0,0 +1,238 @@
+		opensearchRam = 32
+		osNodeCount = int(math.Ceil(float64(totalDataWithReplicationGB) / float64(opensearchRam) / MemoryDiskRatio))
+	}
+	if osNodeCount > 5 {
+		opensearchRam = 64
+		osNodeCount = int(math.Ceil(float64(totalDataWithReplicationGB) / float64(opensearchRam) / MemoryDiskRatio))
+	}
+
+	if osNodeCount%2 == 0 {
+		osNodeCount++
+	}
+	if osNodeCount < 3 {
+		osNodeCount = 3
+	}
+
+	IndexingOverhead := 0.1
+	DesiredShardSize := 50.0
+	ShardCount := int(float64(totalDataWithReplicationGB) * (1 + IndexingOverhead) / DesiredShardSize)
+
+	MaxShardCount := int(osNodeCount) * (opensearchRam / 2) * 20
+	if ShardCount > MaxShardCount {
+		return 0, 0, 0, errors.New(fmt.Sprint("MaxShardCount:", MaxShardCount, " is less than Shard Count: ", ShardCount))
+	}
+
+	if osNodeCount > 9 {
+		return 0, 0, 0, errors.New(fmt.Sprint("opensearch node count needed: ", osNodeCount, ", please see if you can reduce the input requirements so the OpenSearch Nodes needed comes down to max 9 for 64gb ram"))
+	}
+
+	diskWatermarkThreshold := 0.15
+	marginOfError := 0.1
+	totalStorageNeed := int(float64(totalDataWithReplicationGB) * (1 + diskWatermarkThreshold + marginOfError))
+	osNodeStorage = int(math.Ceil(float64(totalStorageNeed) / float64(osNodeCount)))
+	osNodeStorage = int(math.Round(float64(osNodeStorage)/100.0) * 100.0)
+
+	return
+}
+
+func (h *HardwareCalReq) PostgresqlCal() (pgNodeRam, pgStorage int) {
+	pgStorageNeedGB := (h.NoOfNodes * 20) / 1024
+	pgNodeRam = int(float64(pgStorageNeedGB) * 0.02)
+	if pgNodeRam < 8 {
+		pgNodeRam = 8
+	}
+	pgStorage = 200
+	if pgStorageNeedGB > pgStorage {
+		pgStorage = int(math.Round(float64(pgStorageNeedGB)/100.0) * 100.0)
+	}
+	return
+}
+
+func (h *HardwareCalReq) FrontendCal() (frontendNodeCount, frontendRam int) {
+	concurrentComp := h.NoOfNodes
+	if h.FreqCountComplianceScanPerDay == 0 {
+		concurrentComp = 0
+	}
+	concurrentClient := h.NoOfNodes
+	if h.FreqCountClientRunPerDay == 0 {
+		concurrentClient = 0
+	}
+	concurrentEvent := h.NoOfNodes
+	if h.FreqCountEventFeedPerDay == 0 {
+		concurrentEvent = 0
+	}
+	concurrencyNeeded := concurrentComp + concurrentClient + concurrentEvent
+
+	frontendNodeCount = 2
+	frontendRam = 8
+	processPerMB := 5
+	processesPerGB := processPerMB * 1024
+	_ = h.checkFrontendNodeCount(&frontendNodeCount, frontendRam, processesPerGB, concurrencyNeeded)
+
+	if frontendNodeCount > 3 {
+		frontendNodeCount = 2
+		frontendRam = 16
+		_ = h.checkFrontendNodeCount(&frontendNodeCount, frontendRam, processesPerGB, concurrencyNeeded)
+	}
+	if frontendNodeCount > 3 {
+		frontendNodeCount = 2
+		frontendRam = 32
+		_ = h.checkFrontendNodeCount(&frontendNodeCount, frontendRam, processesPerGB, concurrencyNeeded)
+	}
+	return
+}
+
+func (h *HardwareCalReq) checkFrontendNodeCount(frontendNodeCount *int, frontendRam, processesPerGB, concurrencyNeeded int) (maxConcurrency int) {
+	maxConcurrency = *frontendNodeCount * (frontendRam * processesPerGB)
+	for i := 0; i < 20; i++ {
```

Comment:
Why are we running this loop fixed number of times?

Reply:
to limit max no. of frontend nodes to max 20 for each frontend. But we can remove it for last machine type of 32. But I felt that 20 seems large count for Automate and Chef Server each.